Runtime option to disable OvS offload to P4 target

n-sandeep opened the pull request:
Control OVS offload with an environment variable during runtime. If env variable OVS_P4_OFFLOAD=false, then disable OVS offload, else if OVS_P4_OFFLOAD is not set or OVS_P4_OFFLOAD is any value other than false, then by default enable OVS offload.

n-sandeep committed:
Runtime option to disable OvS offload to P4 target

Control OVS offload with an environment variable during runtime.
If env variable OVS_P4_OFFLOAD=false, then disable OVS offload, else
if OVS_P4_OFFLOAD is not set or OVS_P4_OFFLOAD is any value other
than false, then by default enable OVS offload.

Signed-off-by: Sandeep N <sandeep.nagapattinam@intel.com>

ffoulkes commented on include/openvswitch/ovs-p4rt.h:
```diff
@@ -15,6 +15,15 @@
 extern "C" {
 #endif
 
+/* Control OVS offload with an environment variable during runtime.
```
This is the wrong place to add this definition.

`ovs-p4rt.h` defines the public interface between OvS and `ovs-p4rt`. (It says so at the top of the file.)

The environment variable is part of OvS, not something exported by ovs-p4rt. Neither the environment variable name nor the code that queries it belongs in the interface header file.

n-sandeep replied:
Done agreed Derek, will create a new header file specifically for P4 use cases in OvS

ffoulkes commented on include/openvswitch/ovs-p4rt.h:
```diff
@@ -15,6 +15,15 @@
+ * if OVS_P4_OFFLOAD is not set or OVS_P4_OFFLOAD is any value other
+ * than false, then by default enable OVS offload.
+ */
+#define ENABLE_OVS_P4_OFFLOAD                                                     \
```
This is misleading. A bare #define is assumed to be a static value. This isn't a static value, it's a dynamic value.

My recommendation would be to define a static inline function in a different header file (maybe `p4ovs.h`?).

```c
static inline bool ovs_p4_offload_enabled(void) {
  const char* offload = getenv("OVS_P4_OFFLOAD");
  return (offload == NULL) || strcmp(offload, "false") != 0;
}
```

n-sandeep replied:
Sure, can add an inline function too.

ffoulkes commented on lib/mac-learning.c:
```diff
@@ -616,12 +616,14 @@ mac_learning_expire(struct mac_learning *ml, struct mac_entry *e)
-    fdb_info.is_vlan = true;
-    fdb_info.bridge_id = ml->p4_bridge_id;
-    ConfigFdbTableEntry(fdb_info, false);
+    if (ENABLE_OVS_P4_OFFLOAD) {
```
```suggestion
    if (ovs_p4_offload_enabled()) {
```

ffoulkes commented on lib/mac-learning.c:
```diff
@@ -616,12 +616,14 @@ mac_learning_expire(struct mac_learning *ml, struct mac_entry *e)
-    ConfigFdbTableEntry(fdb_info, false);
+    if (ENABLE_OVS_P4_OFFLOAD) {
+        struct mac_learning_info fdb_info;
+        memset(&fdb_info, 0, sizeof(struct mac_learning_info));
```
```suggestion
        memset(&fdb_info, 0, sizeof(fdb_info));
```

n-sandeep replied:
done, resolved

ffoulkes commented on ofproto/ofproto-dpif-xlate.c:
```diff
@@ -3249,13 +3249,14 @@ xlate_normal(struct xlate_ctx *ctx)
-    } else {
-        VLOG_DBG("Error retrieving FDB information, skipping programming "
-                 "P4 entry");
+    if (ENABLE_OVS_P4_OFFLOAD) {
```
```suggestion
    if (ovs_p4_offload_enabled()) {
```

ffoulkes commented on ofproto/ofproto-dpif-xlate.c:
```diff
@@ -8527,18 +8528,20 @@ xlate_add_static_mac_entry(const struct ofproto_dpif *ofproto,
-    } else {
-        VLOG_DBG("Error retrieving FDB information, skipping programming "
-                 "P4 entry");
+    if (ENABLE_OVS_P4_OFFLOAD) {
```
```suggestion
    if (ovs_p4_offload_enabled()) {
```

ffoulkes commented on ofproto/ofproto-dpif-xlate.c:
```diff
@@ -8527,18 +8528,20 @@ xlate_add_static_mac_entry(const struct ofproto_dpif *ofproto,
-                 "P4 entry");
+    if (ENABLE_OVS_P4_OFFLOAD) {
+        /* Static MAC is configured, program P4 forwarding table */
+        struct xport *ovs_port = get_ofp_port(xbundle->xbridge,
```
Check for null return value?

n-sandeep replied:
if ovs_port  is NULL, this API get_fdb_data will return -1 and that will lead into else case, and we LOG a message

n-sandeep committed:
Addressing review comments and fixing UT defects

This change introduces a mutex and this mutex is taken before
making a p4runtime call to program FDB entry. This is to avoid multiple
parallel calls when OvS creates multiple revalidator and handler threads
which tries to make a p4runtime call when a MAC is learnt.

Signed-off-by: Sandeep N <sandeep.nagapattinam@intel.com>

ffoulkes reviewed (approved):
LGTM

5abeel reviewed (approved):
Please make sure that the documentation is updated to capture this environment variable.

Rest LGTM
